File: 11-性能分析/03-基准测试方法论/统计检验的选择/go/test_selection.go
// 统计检验的选择(Go 版) —— 场景实证与断言; 统计原语在 stats_core.go。
//
// 场景(与 python/test_selection.py 一一对应):
//   A 整组平移      -> t 与 U 都显著
//   B 局部抬高 8/40 -> t 显著(均值被拉动)、U 不显著(秩几乎不变)
//   C 重尾          -> 散布次序 mean >> winsorized > trimmed > median
//   D 效率          -> U 的功效逼近 t 的 3/π(渐近相对效率), 有限样本会偏离
//   E 覆盖率        -> 强偏态下均值 t 区间失真, 中位数次序统计量区间保持有效
//   F Cauchy        -> 样本量增加并不改善均值估计
package main

import (
	"fmt"
	"math"
	"math/rand"
)

func stdev(v []float64) float64 { return math.Sqrt(variance(v)) }

// deterministicGroups: B 是等间距序列, A 与 B 相同但每 5 个样本被抬高到 140。
func deterministicGroups(n int) ([]float64, []float64) {
	b := make([]float64, n)
	for i := range b {
		b[i] = 100 + (float64(i)-float64(n-1)/2)*0.25
	}
	a := append([]float64(nil), b...)
	for i := 0; i < n; i += 5 {
		a[i] = 140
	}
	return a, b
}

func synthLognormal(rng *rand.Rand, n int, sigma float64) []float64 {
	out := make([]float64, n)
	for i := range out {
		out[i] = math.Exp(rng.NormFloat64() * sigma)
	}
	return out
}

func main() {
	// 1) t 分布锚点
	tc := tCritical(10, alpha)
	if !closeTo(tc, 2.228139, 1e-5) || !closeTo(tTwoSidedP(2.228139, 10), 0.05, 1e-6) {
		panic("t distribution anchors broken")
	}
	fmt.Printf("[锚点] t 分布 df=10 双尾临界值 %.6f(教科书 2.228139)\n", tc)

	// 2) NIST 7.3.1 的 Welch 算例
	m1, sd1, n1 := 36.0909, 4.9082, 11.0
	m2, sd2, n2 := 32.2222, 2.5386, 9.0
	v1, v2 := sd1*sd1/n1, sd2*sd2/n2
	t := (m1 - m2) / math.Sqrt(v1+v2)
	df := (v1 + v2) * (v1 + v2) / (v1*v1/(n1-1) + v2*v2/(n2-1))
	if !closeTo(t, 2.2694, 2e-3) || df <= 15.4 || df >= 15.7 {
		panic(fmt.Sprintf("NIST example mismatch: t=%v df=%v", t, df))
	}
	fmt.Printf("[锚点] NIST 7.3.1 算例: t=%.4f(手册 2.2694), ν=%.2f(手册 15.5)\n", t, df)

	// 3) 场景 A / B
	a, b := deterministicGroups(40)
	shifted := make([]float64, len(b))
	for i, v := range b {
		shifted[i] = v + 3
	}
	_, _, pA := welchT(b, shifted)
	_, pUA := mannWhitneyP(b, shifted)
	if pA >= 1e-3 || pUA >= 1e-3 {
		panic("scenario A: both tests should be significant")
	}
	_, _, ptB := welchT(a, b)
	_, puB := mannWhitneyP(a, b)
	if ptB >= 0.05 || puB <= 0.05 {
		panic(fmt.Sprintf("scenario B mismatch: t=%v U=%v", ptB, puB))
	}
	fmt.Printf("[场景A 整组平移] t p=%.2e U p=%.2e(都显著)\n", pA, pUA)
	fmt.Printf("[场景B 局部抬高] t p=%.4f(显著) U p=%.4f(不显著) -> 同一批数据两种结论\n", ptB, puB)

	// 4) 场景 C: 位置估计量稳定性
	rng := rand.New(rand.NewSource(7))
	estimators := map[string]func([]float64) float64{
		"mean": mean, "median": median,
		"trimmed5%":    func(v []float64) float64 { return trimmedMean(v, 0.05) },
		"winsorized5%": func(v []float64) float64 { return winsorizedMean(v, 0.05) },
	}
	spread := map[string]float64{}
	for name, fn := range estimators {
		vals := make([]float64, 300)
		for i := range vals {
			s := synthLognormal(rng, 21, 1.5)
			s[rng.Intn(21)] = s[0] * 1000 // 单个 1000 倍离群点
			vals[i] = fn(s)
		}
		spread[name] = stdev(vals)
	}
	if !(spread["mean"] > 3*spread["median"] &&
		spread["trimmed5%"] < spread["mean"] && spread["winsorized5%"] < spread["mean"]) {
		panic(fmt.Sprintf("scenario C mismatch: %v", spread))
	}
	fmt.Printf("[场景C 重尾] 散布 mean=%.3f winsorized5%%=%.3f trimmed5%%=%.3f median=%.3f\n",
		spread["mean"], spread["winsorized5%"], spread["trimmed5%"], spread["median"])

	// 5) 场景 D: U 相对 t 的效率(ARE 是渐近量, 有限样本功效比会偏离)
	for _, cfg := range []struct {
		n      int
		effect float64
		wantLo float64
	}{{10, 0.6, 0.85}, {150, 0.25, 0.93}} {
		hitT, hitU, trials := 0, 0, 2000
		for i := 0; i < trials; i++ {
			s1 := make([]float64, cfg.n)
			s2 := make([]float64, cfg.n)
			for j := 0; j < cfg.n; j++ {
				s1[j] = rng.NormFloat64()
				s2[j] = rng.NormFloat64() + cfg.effect
			}
			if _, _, p := welchT(s1, s2); p < alpha {
				hitT++
			}
			if _, p := mannWhitneyP(s1, s2); p < alpha {
				hitU++
			}
		}
		ratio := float64(hitU) / float64(hitT)
		if ratio < cfg.wantLo || ratio > 1.02 {
			panic(fmt.Sprintf("scenario D mismatch n=%d ratio=%v", cfg.n, ratio))
		}
		fmt.Printf("[场景D 效率] n=%-4d 功效 t=%.3f U=%.3f -> 比值 %.3f\n",
			cfg.n, float64(hitT)/float64(trials), float64(hitU)/float64(trials), ratio)
	}
	fmt.Printf("     ARE 理论 3/π ≈ %.3f: 有限样本偏离, 大样本才收敛\n", 3/math.Pi)

	// 6) 场景 E: 覆盖率
	covT, covMed, trials := 0, 0, 600
	trueMean, trueMed := math.Exp(1.5*1.5/2), 1.0
	for i := 0; i < trials; i++ {
		s := synthLognormal(rng, 10, 1.5)
		if lo, hi := meanCIT(s, alpha); lo <= trueMean && trueMean <= hi {
			covT++
		}
		if mlo, mhi := medianCIOrderStatistic(s, alpha); mlo <= trueMed && trueMed <= mhi {
			covMed++
		}
	}
	ct, cm := float64(covT)/float64(trials), float64(covMed)/float64(trials)
	if ct >= 0.92 || cm < 0.90 {
		panic(fmt.Sprintf("scenario E mismatch: t=%v med=%v", ct, cm))
	}
	fmt.Printf("[场景E 覆盖率] 对数正态(σ=1.5,n=10): 均值 t 区间 %.3f(名义 0.95), 中位数区间 %.3f\n", ct, cm)

	// 7) 场景 F: Cauchy
	spreads := []float64{}
	for _, n := range []int{10, 1000} {
		vals := make([]float64, 120)
		for i := range vals {
			s := make([]float64, n)
			for j := range s {
				s[j] = rng.NormFloat64() / rng.NormFloat64()
			}
			vals[i] = mean(s)
		}
		spreads = append(spreads, stdev(vals))
	}
	meds := make([]float64, 120)
	for i := range meds {
		s := make([]float64, 1000)
		for j := range s {
			s[j] = rng.NormFloat64() / rng.NormFloat64()
		}
		meds[i] = median(s)
	}
	if !(spreads[1] > 0.5*spreads[0]) || stdev(meds) >= 0.3*spreads[1] {
		panic("scenario F mismatch")
	}
	fmt.Printf("[场景F Cauchy] 均值散布 n=10 -> %.2f, n=1000 -> %.2f(不下降); 中位数(n=1000) -> %.2f\n",
		spreads[0], spreads[1], stdev(meds))

	fmt.Println("\ntest_selection: 全部自检通过")
}

File: 11-性能分析/03-基准测试方法论/统计检验的选择/python/test_selection.py
#!/usr/bin/env python3
"""统计检验的选择: 什么时候用均值 ± t 区间, 什么时候必须用中位数 + 分位数/秩检验。

结论先行(与 README 一一对应):
  * t 检验回答"两组**均值**是否不同", 前提是独立同分布、小样本时还要近似正态;
    重尾下均值本身不稳定 —— NIST 明确写"均值缺乏 robustness of validity", 而 Cauchy
    分布下"收集更多数据并不能更准地估计均值"。
  * Mann-Whitney U 回答"一组是否**随机大于**另一组"(P(X>Y) != 0.5), 不做分布假设,
    代价是对正态数据损失效率(ARE = 3/π ≈ 0.955)。
  * 两者回答不同问题, 所以会出现"t 显著、U 不显著"这种看似矛盾的结果 —— 选检验
    的前提是先决定你要回答哪个问题。

统计原语在 stats_core.py, 本文件只做场景实证与断言。
"""

from __future__ import annotations

import math
import random
import statistics
from typing import List, Tuple

from stats_core import (ALPHA, mann_whitney_p, mean_ci_t, median_ci_order_statistic,
                        t_critical, t_two_sided_p, trimmed_mean, welch_t, winsorized_mean)


def _deterministic_groups(n: int = 40) -> Tuple[List[float], List[float]]:
    """无随机性的两组: B 是等间距序列, A 与 B 相同但每 5 个样本被抬高到 140。"""
    b = [100 + (i - (n - 1) / 2) * 0.25 for i in range(n)]
    a = list(b)
    for idx in range(0, n, 5):
        a[idx] = 140.0
    return a, b


def _self_test() -> None:
    # 1) t 分布锚点
    tc = t_critical(10, 0.05)
    assert abs(tc - 2.228139) < 1e-5, tc
    assert abs(t_two_sided_p(2.228139, 10) - 0.05) < 1e-6
    assert abs(t_two_sided_p(1.96, 1e7) - 0.05) < 2e-3  # 大自由度退化到正态
    print(f"[锚点] t 分布 df=10 双尾临界值 {tc:.6f}(教科书 2.228139)")

    # 2) 复现 NIST 手册 7.3.1 的 Welch 算例
    m1, sd1, n1 = 36.0909, 4.9082, 11
    m2, sd2, n2 = 32.2222, 2.5386, 9
    v1, v2 = sd1 ** 2 / n1, sd2 ** 2 / n2
    t = (m1 - m2) / math.sqrt(v1 + v2)
    df = (v1 + v2) ** 2 / (v1 ** 2 / (n1 - 1) + v2 ** 2 / (n2 - 1))
    assert abs(t - 2.2694) < 2e-3, t
    assert 15.4 < df < 15.7, df
    print(f"[锚点] NIST 7.3.1 算例: t={t:.4f}(手册 2.2694), ν={df:.2f}(手册 15.5)")

    # 3) 场景 A: 整组平移(真实均值差异) -> 两种检验都显著
    a, b = _deterministic_groups()
    shifted = [v + 3.0 for v in b]
    p_t_a = welch_t(b, shifted)[2]
    p_u_a = mann_whitney_p(b, shifted)[1]
    assert p_t_a < 1e-3 and p_u_a < 1e-3, (p_t_a, p_u_a)
    print(f"[场景A 整组平移] t p={p_t_a:.2e}, U p={p_u_a:.2e} -> 都显著")

    # 4) 场景 B: 8/40 个样本被抬高 40 -> 均值被拉动(t 显著), 秩几乎不变(U 不显著)
    p_t_b = welch_t(a, b)[2]
    p_u_b = mann_whitney_p(a, b)[1]
    assert p_t_b < 0.05 and p_u_b > 0.05, (p_t_b, p_u_b)
    print(f"[场景B 局部抬高] t p={p_t_b:.4f}(显著), U p={p_u_b:.4f}(不显著)")
    print("     同一批数据两种结论 —— t 检测均值, U 检测随机优势(P(X>Y)-0.5)")

    # 5) 场景 C: 重尾下各位置估计量的稳定性, 以及单个极端值对均值/中位数的影响
    rng = random.Random(7)
    spikes = [math.exp(rng.gauss(0.0, 1.5)) for _ in range(21)]
    spikes[0] = spikes[0] * 1000
    reps = 300
    estimators = {"mean": statistics.fmean, "median": statistics.median,
                  "trimmed5%": trimmed_mean, "winsorized5%": winsorized_mean}
    spread = {}
    for name, fn in estimators.items():
        vals = []
        for _ in range(reps):
            s = [math.exp(rng.gauss(0.0, 1.5)) for _ in range(21)]
            s[rng.randrange(21)] = s[0] * 1000
            vals.append(fn(s))
        spread[name] = statistics.stdev(vals)
    assert spread["mean"] > 3 * spread["median"], spread
    assert spread["trimmed5%"] < spread["mean"] and spread["winsorized5%"] < spread["mean"]
    print("[场景C 重尾] 位置估计量散布(越小越稳): " +
          ", ".join(f"{k}={v:.3f}" for k, v in sorted(spread.items(), key=lambda kv: kv[1])))
    print(f"     单个 1000 倍离群点: 均值 {statistics.fmean(spikes):.1f} vs 中位数 "
          f"{statistics.median(spikes):.3f}(真值 ≈1)")

    # 6) 场景 D: 正态下 U 相对 t 的效率。ARE 是 n→∞ 的极限, 有限样本功效比不等于它
    ratios = {}
    for n, effect, trials in ((10, 0.6, 3000), (150, 0.25, 2000)):
        hit_t = hit_u = 0
        for _ in range(trials):
            s1 = [rng.gauss(0, 1) for _ in range(n)]
            s2 = [rng.gauss(effect, 1) for _ in range(n)]
            hit_t += 1 if welch_t(s1, s2)[2] < ALPHA else 0
            hit_u += 1 if mann_whitney_p(s1, s2)[1] < ALPHA else 0
        ratios[n] = (hit_t / trials, hit_u / trials, hit_u / hit_t)
    assert 0.85 <= ratios[10][2] <= 1.02, ratios[10]
    assert ratios[150][2] >= 0.93, ratios[150]
    for n in (10, 150):
        pt, pu, r = ratios[n]
        print(f"[场景D 效率] n={n:<4} 功效 t={pt:.3f} U={pu:.3f} -> 比值 {r:.3f}")
    print(f"     ARE 理论 3/π ≈ {3 / math.pi:.3f}: n=10 的 {ratios[10][2]:.3f} 明显偏离, "
          f"n=150 的 {ratios[150][2]:.3f} 才靠近")

    # 7) 场景 E: 强偏态下均值 t 区间覆盖率失真, 中位数区间保持有效
    cov_t = cov_med = 0
    trials_e = 600
    for _ in range(trials_e):
        s = [math.exp(rng.gauss(0.0, 1.5)) for _ in range(10)]
        true_mean, true_med = math.exp(1.5 ** 2 / 2), 1.0
        lo, hi = mean_ci_t(s)
        cov_t += 1 if lo <= true_mean <= hi else 0
        mlo, mhi = median_ci_order_statistic(s)
        cov_med += 1 if mlo <= true_med <= mhi else 0
    cov_t, cov_med = cov_t / trials_e, cov_med / trials_e
    assert cov_t < 0.92, cov_t
    assert cov_med >= 0.90, cov_med
    print(f"[场景E 覆盖率] 对数正态(σ=1.5, n=10): 均值 t 区间 {cov_t:.3f}(名义 0.95), "
          f"中位数区间 {cov_med:.3f}")

    # 8) 场景 F: Cauchy —— 加样本量也不能改善均值估计(NIST 原文结论)
    spreads = []
    for n in (10, 1000):
        vals = [statistics.fmean([rng.gauss(0, 1) / rng.gauss(0, 1) for _ in range(n)])
                for _ in range(120)]
        spreads.append(statistics.stdev(vals))
    med_vals = [statistics.median([rng.gauss(0, 1) / rng.gauss(0, 1) for _ in range(1000)])
                for _ in range(120)]
    assert spreads[1] > 0.5 * spreads[0], spreads
    assert statistics.stdev(med_vals) < 0.3 * spreads[1]
    print(f"[场景F Cauchy] 均值散布 n=10 -> {spreads[0]:.2f}, n=1000 -> {spreads[1]:.2f}"
          f"(不下降); 中位数(n=1000) -> {statistics.stdev(med_vals):.2f}")


if __name__ == "__main__":
    _self_test()
    print("\ntest_selection: 全部自检通过")
